README.md
# Chrome Logger Firefox WebExtension

WebExtension implementing [Chrome Logger protocol](https://craig.is/writing/chrome-logger/techspecs) for Firefox.

## String Substitutions And Styling

Console output is automatically styled from settings *unless* non-escaped [string substitutions](https://developer.mozilla.org/en-US/docs/Web/API/console#Outputting_text_to_the_console) are detected in the message being logged.

## Logging Data From DOM

To log data to the web console through this extension *after* headers have been sent, the following methods are supported.

Log the entire ChromeLoggerData object that would've otherwise been the base64-encoded header value.

	<script 
		id="chromelogger" 
		type="application/json"
	>{ ChromeLoggerData object }</script>	

Or just log the `rows` array of the ChromeLoggerData object. Attributes `data-chromelogger-version` and `data-chromelogger-columns` are optional and default to the values shown if missing.

	<script 
		id="chromelogger" 
		type="application/json" 
		data-chromelogger-version="2.0" 
		data-chromelogger-columns="log,backtrace,type"
	>[ ChromeLoggerData.rows array ]</script>

`<script>` values must be valid JSON or error is thrown.

CHANGELOG.md
# Chrome Logger Firefox WebExtension Change Log

### 2019 Nov 24 / Version 2.0
* Removed: `<script[data-chromelogger-data]>` and `<script[data-chromelogger-rows]>` node parsing.
* Added: `<script[type="application/json"]#chromelogger>` node parsing.
* Updated: moved DOM parsing functionality from devtools eval'd code (Mozilla no likey!) to injected content script.
* Fixed: dark, transparent icon did not display well w/dark theme.

### 2018 Oct 09 / Version 1.7.3
* Fixed: null "object" resulted in error when checked for property, ending processing of rest of data stack.

### 2018 Aug 06 / Version 1.7.2
* Fixed: invalid `data.rows[...].type` values resulted in `console` error.

### 2018 Aug 04 / Version 1.7.1
* Fixed: broken `table` and various other `console` methods.
* Fixed: inadequate substitution pattern detection and subsequent unescaping.

### 2018 Aug 03 / Version 1.7
* Fixed: columns arrays that were arranged in ways or [in|ex]cluded columns other than what was explicitly specified by the protocol failed to process. @shoutout Zxurian
* Added: Support for `[object].___class_name` and Class Name substitution styling.
* Added: Support for `X-ChromePHP-Data` headers.
* Updated: Now scans headers from all resource types.
* Updated: `data-chromelogger-version="1.0"` and `data-chromelogger-columns="log,backtrace,type"` optional attributes parsed from `<script[data-chromelogger-rows]>` nodes.
* Added: `<script[data-chromelogger-data]>` node parsing.
* Added: `display_data_url=[bool]` setting and Data Header substitution styling.
* Updated: removed "METHOD url" display from being appended to first parsed data row, logged separately if `display_data_url=true`.

### 2018 Apr 24 / Version 1.6.1
* Fixed: trying (and failing, of course) to process null values as objects w/enumerated properties.
* Fixed: tabs.executeScript() fails injecting both file + code, throws to devtools fallback.

### 2018 Apr 23 / Version 1.6
* Restored: removal of `__proto__` and `length` properties from objects and arrays.

### 2018 Mar 26 / Version 1.5
* Added: console message styling for various contexts w/options page to customize.
* Added: parsing of ChromeLogger formatted rows data from loaded DOM content. Syntax: `<script data-chromelogger-rows="{key}">var {key} = [[log, backtrace, type], ...];</script>`
* Issue: cannot retrieve rows data from subframes DOM until Firefox supports devtools.inspectedWindow.eval() options parameter!
* Regression: `__proto__` and `length` object and array properties no longer removed. Problematic.

### 2018 Mar 18 / Version 1.4
* Updated: chromelogger protocol trace value now appended to passed console method arguments. @shoutout Erik

### 2017 Dec 23 / Version 1.3
* Added: `__proto__` and `length` properties are removed from objects and arrays respectively output to web console.

### 2017 Dec 16 / Version 1.2
* Added: Tab.fallback, flagged true if log.js fails to inject onDOMReady as in the case when Firefox loads JSON viewer in the tab inspectedWindow.
* Updated: Tab.log() sends parsed ChromeLoggerData back to dev.js if Tab.fallback is true.
* Updated: dev.js runtime.Port.onMessage handler stringifies and runs ChromeLoggerData from Tab.log() through devtools.inspectedWindow.eval(), dirty but effective.

### 2017 Nov 29 / Version 1.1.1
* Updated: Chrome Logger data row values validation, replacement on missing with defaults. @shoutout Honza
* Fixed: warning on icon being wrong size.

### 2017 Nov 19 / Version 1.1
* Added: webNavigation event handlers to replace spotty tab.onUpdated event handler.
* Updated: Chrome Logger data rows callback catches malformed row data.

### 2017 Nov 17 / Version 1.0.1
* Fixed: was matching against case sensitive X-ChromeLogger-Data header. Now case insensitive match. @shoutout Jan

### 2017 Oct 10 / Version 1.0
* Maiden voyage.

dev.js
"use strict";
/**
 * DevTools
 * Triggers the parsing, processing and outputting to the web console messages encoded in X-ChromeLogger-Data headers.
 */

/**
 * Connects the tab these devtools are open on to the background and content scripts via runtime.onConnect
 * @since 1.0
 * @var runtime.Port
 */
var port = browser.runtime.connect({
	// this is used as a tab key to mark open console sessions and derive the tabId from ...
	name: 'tab' + browser.devtools.inspectedWindow.tabId
});


/**
 * Listener / In the unlikely event of disconnect from the background process.
 * @since 1.0
 * @param runtime.Port
 */
port.onDisconnect.addListener(( port ) => {
	if ( port.error ) console.error( 'Disconnected due to error: %s', port.error.message );
	else console.error( '%s has been disconnected!', port.name );
});


/**
 * Listener / Passed details from Tab.onHeadersReceived() handler, passes them back to onDevPortMessage() for processing if the devtools here are open.
 *
 * @since 1.0
 * @since 1.2
 *	- distinguish between tabs.onHeadersReceived and ChromeLoggerData
 *	- convert ChromeLoggerData into script and eval against inspectedWindow
 * @since 1.5
 *	- moved all data processing to bg.js onDevPortMessage()
 *	- updated to use data.args processed from data.rows
 * @since 1.7
 *	- parse <script[data-chromelogger-rows]> nodes data-chromelogger-version and data-chromelogger-columns attributes
 *	- parse <script[data-chromelogger-data]> nodes.
 * @since 2.0
 *	- removed details.processContentUrl processing, moved to log.js
 *
 * @param tabs.onHeadersReceived|ChromeLoggerData details
 */
port.onMessage.addListener(( details ) => {

	// details contains response headers ? bounce back to background script for processing ...
	if ( details.responseHeaders ) port.postMessage( details );

	// ChromeLoggerData ? sent here if log.js fails to inject into the tab ...
	else if ( details.args ) {

		browser.devtools.inspectedWindow.eval(
			'(function(){ '
				+ cleanObjectProperties.toString() + ' '
				+ JSON.stringify(details.args) + '.forEach(args=>{ '
					+ 'try { '
						+ 'console[ args.shift() ].apply(null, args.map(cleanObjectProperties)); '
					+ '} catch ( error ) { console.error(error); } '
				+ '}); '
				+ 'return true; '
			+ '})();'
		).then(([ success, failure ])=>{
			if ( failure ) console.error(failure);
		});

	}

});


log.js
"use strict";
/**
 * Injection script.
 *
 * @since 1.0
 * @since 1.5
 *	- moved all data processing to bg.js onDevPortMessage()
 *	- updated to use data.args processed from data.rows
 * @since 2.0
 *	- tabId is passed upon initialization from bg.js.
 *
 * @param object details
 *	- processed ChromeLogger data sent from Tab.log() or tabId upon initialization
 */
browser.runtime.onMessage.addListener(( details )=>{

	// has args ? processed chromelogger data, output to console ...
	if ( details.args ) {

		details.args.forEach(args=>{
			try {
				window.console[ args.shift() ].apply(null, args.map(cleanObjectProperties));
			} catch ( error ) {
				window.console.error( error, args );
			}
		});

	}

	// has tabId ? parse any chromelogger data from document ...
	else if ( details.tabId ) {

		// array of ChromeLoggerData objects ...
		let data = [];

		// check for old DOM elements, issue notice if found ...
		if ( document.querySelector('script[data-chromelogger-data],script[data-chromelogger-rows]') ) data.push({
			columns: [ 'type', 'log' ],
			rows: [[ 'info', [
				"Support for parsing %c<script data-chromelogger-data>%c and %c<script data-chromelogger-rows>%c nodes from DOM has been removed from %s %s.\n\nPlease upgrade to %c<script id=\"chromelogger\" type=\"application/json\">%c nodes instead.\n\nSee %s for details.",
				'font-family:monospace;color:red;',
				'font-family:unset;color:unset;',
				'font-family:monospace;color:red;',
				'font-family:unset;color:unset;',
				browser.runtime.getManifest().name,
				browser.runtime.getManifest().version,
				'font-family:monospace;color:dodgerblue;',
				'font-family:unset;color:unset;',
				browser.runtime.getManifest().homepage_url
			] ]],
			version: '2.0'
		});

		// parse #chromelogger JSON script blocks ...
		document.querySelectorAll('script[type^="application/json"]#chromelogger').forEach(node=>{

			try {

				// parse JSON or throw error ...
				let value = JSON.parse(node.innerText);

				// array ? format rows into chromelogger data object ...
				if (
					Array.isArray(value)
				) data.push({
					rows: value,
					columns: (
						( value = node.dataset.chromeloggerColumns )
						? value.replace(/\s/,'').split(',')
						: [ 'log', 'backtrace', 'type' ]
					),
					version: (
						( value = node.dataset.chromeloggerVersion )
						? value
						: '2.0'
					)
				});

				// object w/rows array ? chromelogger data object ...
				else if (
					typeof value == 'object'
					&& value.rows
					&& Array.isArray(value.rows)
				) data.push(value);

				// fail ! invalid data ...
				else window.console.error('Invalid ChromeLogger Data!', value);

			} catch ( error ) {
				window.console.error(error, node.innerText);
			}

		});

		// send back any data to process ...
		data.forEach(data=>{
			data.tabId = details.tabId;
			browser.runtime.sendMessage(data).catch(error=>{ console.error(error); });
		});

	}

});


global.js
"use strict";
/**
 * Global scripts.
 */

/**
 * Default options object.
 * @since 1.5
 * @since 1.7
 *	- added classname substitution style.
 *	- added display_data_url option.
 * @var object
 */
const DEFAULT_OPTIONS = {
	console_substitution_styles: {
		error: 'color:red;',
		warn: 'color:orange;',
		info: 'color:limegreen;',
		log: '',
		group: 'color:mediumturquoise;border-bottom:1px dashed;cursor:pointer;',
		number: 'background-color:dodgerblue;color:white;font-weight:bold;border-radius:0.5em;padding:0em 0.3em;',
		fileline: 'color:mediumpurple;font-style:italic;border-style:solid;border-width:0px 1px;border-radius:0.5em;padding:0em 0.5em;',
		classname: 'font-weight:bold;',
		header: 'display:block;background-color:black;color:white;text-align:center;padding:0.2em;border-radius:0.3em;'
	},
	display_data_url: true
};


/**
 * Array.map callback
 * Removes "__proto__" and "length" properties from objects and arrays.
 * @since 1.3
 * @param mixed obj
 * @return mixed
  */
function cleanObjectProperties( obj ) {

	// not an object ? return as-is ...
	if ( typeof obj !== 'object' || obj === null ) return obj;

	// removes length property from arrays ...
	if ( Array.isArray(obj) ) obj = Object.assign({}, obj);

	// remove annoying __proto__ property ...
	obj.__proto__ = null;

	// recurse through properties ...
	Object.entries(obj).forEach(([ key, value ])=>{ obj[ key ] = cleanObjectProperties( value ); });

	// return cloned object ...
	return obj;

}


options.js
"use strict";
/**
 * Options page scripts.
 * @since 1.5
 */


/**
 * Populate form from options.
 * @param object opts
 */
function populateForm( opts ) {

	// populate substitution styles ...
	Object.entries(opts.console_substitution_styles).forEach(([key, value])=>{
		let input = document.querySelector( 'input#console_substitution_styles-'.concat(key) );
		if ( input ) {
			input.value = value;
			input.dispatchEvent(new Event('keyup')); // trigger change event to update label
		}
	});

	// misc. settings ...
	document.querySelector('input#display_data_url').checked = opts.display_data_url;

}


document.addEventListener("DOMContentLoaded", event=>{

	// style inputs ...
	let console_substitution_style_inputs = document.querySelectorAll("form fieldset#console_substitution_styles input");

	// update label style on input update ...
	console_substitution_style_inputs.forEach(input=>{
		input.addEventListener("keyup", event=>{
			event.target.previousElementSibling.style = event.target.value;
		});
	});

	// populate form from options ...
	browser.storage.sync.get(DEFAULT_OPTIONS)
	.then(populateForm)
	.catch(error=>{ console.error(error); });

	// onsubmit ...
	document.querySelector("form").addEventListener("submit", event=>{

		event.preventDefault();

		// reset ...
		if ( event.explicitOriginalTarget.id == 'reset' ) {

			// set from default options ? update form from default options ...
			browser.storage.sync.set(DEFAULT_OPTIONS)
			.then(()=>{
				populateForm(DEFAULT_OPTIONS);
			})
			.catch(error=>{ console.error(error); });

		}

		// save ...
		else {

			// styles collection ...
			let console_substitution_styles = {};

			// update collection ...
			console_substitution_style_inputs.forEach(input=>{
				console_substitution_styles[ input.id.substr( input.id.indexOf('-') + 1 ) ] = input.value;
			});

			// save settings ...
			browser.storage.sync.set({

				// styles ...
				console_substitution_styles: console_substitution_styles,

				// display url ...
				display_data_url: document.querySelector("form input#display_data_url").checked,

			})
			.catch(error=>{ console.error(error); });

		}

	});


});



bg.js
"use strict";
/**
 * Background script.
 * All the routine heavy lifting should be done here and the results coordinated betweent dev.js devtools script and log.js content script.
 */

/**
 * Collection of keyed tab objects connected when devtools are opened on a tab.
 * @since 1.0
 * @var object tabs
 */
var tabs = {};


/**
 * Return tab id integer from port.name.
 * @since 1.0
 * @param runtime.Port port
 * @return integer
 */
function tabIdFromPort( port ) {
	return parseInt( port.name.substr(3) );
}


/**
 * Return tabs object key from id integer.
 * @since 1.0
 * @param integer tabId
 * @return string
 */
function tabKeyFromId( tabId ) {
	return 'tab'.concat(tabId);
}


/**
 * Returns a tab object from id.
 * @since 1.0
 * @param integer
 * @return Tab|undefined
 */
function tabFromId( tabId ) {
	return tabs[ tabKeyFromId( tabId ) ];
}


/**
 * Tab class.
 * @since 1.0
 * @since 2.0
 *	- removed processContentUrl() method, functionality moved to log.js
 */
class Tab {

	/**
	 * @since 1.0
	 * @param integer tabId
	 * @param runtime.Port port
	 */
	constructor( tabId, port ){

		/**
		 * Tab id.
		 * @since 1.0
		 * @var integer
		 */
		this.id = tabId;

		/**
		 * Incoming connection from devtools being loaded / set by runtime.onConnect listener.
		 * @since 1.0
		 * @var runtime.Port
		 */
		this.devPort = port;

		/**
		 * Contains pending data while tab main frame is loading.
		 * @since 1.0
		 * @var array
		 */
		this.pending = [];

		/**
		 * Reflects tab DOMReady state. When true the tab is ready to receive messages.
		 * @since 1.1
		 * @var bool
		 */
		this.ready = false;

		/**
		 * Fallback to devtools for reporting should log.js script injection fail onDOMReady.
		 * @since 1.2
		 * @var bool
		 */
		this.fallback = false;

		/**
		 * Sends header details to devtools which, if open, will send them back to onDevPortMessage()
		 * @since 1.0
		 * @param webRequest.onHeadersReceived details
		 * @note This way we're only processing headers when the user can see output from them logged to the web console.
		 * @note Setting up an anonymous function this way in order to make the handler for this tab distinct from that of others so the event handler can tell difference.
		 */
		this.onHeadersReceived = function( details ) {
			tabs[ tabKeyFromId( details.tabId ) ].devPort.postMessage( details );
		}

	}


	/**
	 * Logs data to log.js
	 * @since 1.0
	 * @since 1.2
	 *	- send data back on port if Tab.fallback == true
	 * @param object data
	 *	- ChromeLogger Data
	 */
	log( data ){

		// tab is DOMReady ? send to log.js now !
		if ( this.ready ) browser.tabs.sendMessage( this.id, data );

		// fallback to devtools.inspectedWindow.eval() ? send back through the port now !
		else if ( this.fallback ) this.devPort.postMessage( data );

		// tab document still loading ! queue in pending data to be sent by tabs.onUpdated listener on complete
		else this.pending.push( data );

	}


}


/**
 * Removes a corresponding tab and events / used as tabs.onRemoved handler too.
 * @since 1.0
 * @param integer tabId
 */
function onTabRemoved( tabId ) {

	var tabKey = tabKeyFromId(tabId),
		tab = tabs[ tabKey ];

	if ( tab ) {

		// disconnect ports ...
		tab.devPort.disconnect();

		// remove handler ...
		browser.webRequest.onHeadersReceived.removeListener( tab.onHeadersReceived );

		// remove tab ...
		delete tabs[ tabKey ];

	}

}


/**
 * Process ChromeLogger data rows into console args.
 * @since 1.5
 * @since 1.7
 * 	- flexible support for different row.column arrangments
 * @param ChromeLoggerData object data
 */
function processChromeLoggerData( data ) {

	return new Promise( resolve => {

		// load options, process data, pass to Tab.log() ...
		browser.storage.sync.get(DEFAULT_OPTIONS).then(opts=>{

			// ensure lowercase columns array ...
			data.columns = (
				! data.columns
				|| ! Array.isArray(data.columns)
				? [ 'log', 'backtrace', 'type' ]
				: data.columns.map(column=>{ return column.toLowerCase(); })
			);

			// map to rows array to console method args ...
			data.args = data.rows.map(row=>{

				// convert row to object w/columns mapped to properties ...
				row = row.reduce(function( row, value, index ){
					row[ data.columns[ index ] ] = value;
					return row;
				}, { 'log': [], 'backtrace': false, 'type': 'log' });

				var

				// console method / @see https://developer.mozilla.org/en-US/docs/Web/API/Console
				method = row.type,

				// console.[method] arguments ...
				args = row.log,

				// file:line ...
				fileline = row.backtrace,

				// substitution pattern ...
				tmpl_pattern = '',

				// substition arguments ...
				tmpl_args = [];

				// ensure method is valid ...
				if (
					typeof method != 'string'
					|| ! console[ method ]
				) method = 'log';

				// ensure arguments is array ...
				if ( ! Array.isArray(args) ) args = [ args ];

				// assertion ? ...
				if ( method == 'assert' ) {
					// resolves true ? log nothing ...
					if ( args.shift() ) return false;
					// false ! log error ...
					else method = 'error';
				}

				// process arguments ...
				if (
					args.length > 0
					&& [ 'log', 'info', 'warn', 'error', 'group', 'groupCollapsed' ].includes(method)
				) {

					// detect, passthru an existing substitution pattern ...
					if (
						typeof args[0] == 'string'
						&& /(^|[^%])%(s|d|i|f|o|O|c|\.\d+(d|i|f))/.test(args[0])
					) {
						tmpl_pattern = args.shift();
						tmpl_args = args;
					}

					// generate pattern ...
					else {

						// make array ...
						tmpl_pattern = [];

						// populate pattern and args arrays ...
						args.forEach(arg=>{

							switch ( typeof arg ) {

								case 'string':
									tmpl_pattern.push('%c%s%c');
									tmpl_args.push(
										opts.console_substitution_styles[(
											method == 'groupCollapsed'
											? 'group'
											: method
										)],
										// unescape any passed substitution patterns ...
										arg.replace(/%{2,}(s|d|i|f|o|O|c|\.\d+(d|i|f))/g, '%$1'),
										''
									);
									break;

								case 'number':
									tmpl_pattern.push('%c%s%c');
									tmpl_args.push(opts.console_substitution_styles.number, arg, '');
									break;

								case 'object':

									// resolves to true (not null or undefined) and has special class name property ? prepend and remove ...
									if ( arg && arg.hasOwnProperty('___class_name') ) {
										tmpl_pattern.push('%c%s%c');
										tmpl_args.push(opts.console_substitution_styles.classname, arg.___class_name, '');
										delete arg.___class_name;
									}
									// no break, passthru ...

								default:
									tmpl_pattern.push('%o');
									tmpl_args.push(arg);
									break;

							}

						});

						// stringify pattern ...
						tmpl_pattern = tmpl_pattern.join(' ');

					}

				}

				// straight arguments for all other console methods, no backtrace ...
				else {
					tmpl_args = args;
					fileline = false;
				}

				// append fileline ...
				if ( fileline ) {

					// add a space if there is other pattern content ...
					if ( tmpl_pattern ) tmpl_pattern = tmpl_pattern.concat(' ');

					tmpl_pattern = tmpl_pattern.concat('%c%s');
					tmpl_args.push(opts.console_substitution_styles.fileline, fileline);

				}

				// prepend string pattern to arguments ...
				if ( tmpl_pattern ) tmpl_args.unshift( tmpl_pattern );

				// prepend method ...
				tmpl_args.unshift( method );

				// return processed arguments ...
				return tmpl_args;

			}).filter( args => args !== false );

			// return processed data ...
			resolve(data);

		})
		.catch(console.error);

	});

}


/**
 * Tab.devPort runtime.Port.onMessage event handler.
 * Catches and processes header details sent by Tab.onHeadersReceived() handler to and passed back from a verified open devtools.
 *
 * @since 1.0
 * @since 1.5
 *	- accept rows array retrieved from document by devtools ...
 * @since 1.7
 *	- also checks for X-ChromePHP-Data header
 *	- logs details url, method separately as chromelogger data
 * @since 2.0
 *	- removed capturing details.rows from dev.js
 *
 * @param tabs.onHeadersReceived details
 */
function onDevPortMessage( details ) {

	// details object passed through the open devtools ...
	if ( details.responseHeaders ) {

		// headers to process data from ...
		var headers = [ 'x-chromelogger-data', 'x-chromephp-data' ];

		// parse headers ...
		details.responseHeaders.forEach(( header )=>{

			// ChromeLogger data ! decode and process ...
			if ( headers.includes( header.name.toLowerCase() ) ) {

				// load options, process ...
				browser.storage.sync.get(DEFAULT_OPTIONS).then(opts=>{

					// display data url ? log it as chromelogger data ...
					if ( opts.display_data_url ) processChromeLoggerData({
						rows: [[[
							'%c%s %s',
							opts.console_substitution_styles.header,
							details.method,
							details.url
						]]]
					}).then(data=>{
						tabFromId( details.tabId ).log( data );
					});

					// attempt to parse data from header ...
					try {

						// base64 decode / parse JSON ...
						var data = JSON.parse( atob( header.value ) );

						// process and log ...
						processChromeLoggerData( data ).then(data=>{
							tabFromId( details.tabId ).log( data );
						});

					} catch( error ) { console.error(error); }

				});

			}

		});

	}

}


/**
 * Tab.devPort runtime.Port.onDisconnect event handler.
 * @since 1.0
 * @param runtime.Port port
 */
function onDevPortDisconnect( port ) {

	// report error ...
	if ( port.error ) console.error('Disconnected due to error:', port.error.message);
	else console.error( port.name, 'has been disconnected!' );

	// remove tab if any exists ...
	onTabRemoved( tabIdFromPort(port) );

}


/**
 * Listener / Assigns tab object and handlers connecting devtools context when devtools are opened on a tab.
 *
 * @since 1.0
 * @since 1.7
 *	- removed onHeadersReceived listener "types" to process headers from ALL resources.
 *
 * @param runtime.Port port
 */
browser.runtime.onConnect.addListener(( port ) => {

	var tabId = tabIdFromPort(port),
		tabKey = port.name,
		tab = tabs[ tabKey ];

	// ports change when devtools are closed/opened so [re]assign port event handlers ...
	port.onDisconnect.addListener(onDevPortDisconnect);
	port.onMessage.addListener(onDevPortMessage);

	// no tab ? create it ...
	if ( ! tab ) tab = tabs[ tabKey ] = new Tab( tabId, port );
	// update existing port ...
	else tab.devPort = port;

	// no tab specific anon (IMPORTANT!) listener assigned to catch headers ? assign now ...
	if ( ! browser.webRequest.onHeadersReceived.hasListener( tab.onHeadersReceived ) ) {
		browser.webRequest.onHeadersReceived.addListener(
			tab.onHeadersReceived,
			{
				"urls": [ "<all_urls>" ],
				"tabId": tabId
			},
			[ "responseHeaders" ]
		);
	}

});


/**
 * Listener / receives ChromeLoggerData objects from log.js parsed from document.
 *
 * @since 2.0
 *
 * @param ChromeLoggerData details
 */
browser.runtime.onMessage.addListener(( details )=>{

	processChromeLoggerData( details ).then(data=>{
		tabFromId( details.tabId ).log( data );
	});

});


/**
 * Listener / Assigns onTabRemoved handler for tab removal events.
 * @note Tried to track closing browser.windows events but it doesn't work as of 2017Oct08
 * @todo Whenever a devtools close event becomes available, bind to that instead of this!
 * @since 1.0
 * @param integer tabId
 */
browser.tabs.onRemoved.addListener(onTabRemoved);


/**
 * Reset tab ready state.
 * @since 1.1
 */
browser.webNavigation.onBeforeNavigate.addListener(( details )=>{
	var tab = tabFromId( details.tabId );
	if ( tab && details.frameId == 0 ) {
		tab.ready = false;
		tab.fallback = false;
	}
});


/**
 * Inject log.js and update tab ready state.
 * @since 1.1
 * @since 1.2
 *	- added fallback to devtools reporting if script injection fails
 * @since 1.5
 *	- finally tell tab to process loaded DOM content for additional info to log
 * @since 2.0
 *	- moved DOM content processing from dev.js to log.js
 */
browser.webNavigation.onDOMContentLoaded.addListener(( details )=>{

	var tab = tabFromId( details.tabId );
	if (
		tab
		&& details.frameId == 0 // main tab document ...
	) {

		// inject log.js to receive messages sent to tab ? ...
		browser.tabs.executeScript( details.tabId, { file: '/log.js' })
		.then(()=>{

			// load global functions required by log.js / assume this works at this point ...
			browser.tabs.executeScript( details.tabId, { file: '/global.js' })
			.finally(()=>{

				// update ready state ...
				tab.ready = true;

				// send any pending items ...
				while ( tab.pending.length ) browser.tabs.sendMessage( details.tabId, tab.pending.shift() );

				// tell tab to parse any items from the document itself ...
				browser.tabs.sendMessage( details.tabId, { tabId: details.tabId });

			});

		// unable to inject script ! fallback to devtools functions ...
		}).catch(()=>{

			// fallback to devTools.inspectedWindow.eval() !!!
			tab.fallback = true;

			// send any pending items back through the dev port ...
			while ( tab.pending.length ) tab.devPort.postMessage( tab.pending.shift() );

		});

	}

});
